drivers/CpuUid.h

```diff
@@ -0,0 +1,106 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_CPUUID_H
+#define MBED_CPUUID_H
+
+#include "platform/platform.h"
+
+#if defined(DEVICE_CPUUID) || defined(DOXYGEN_ONLY)
+
+#include <string>
+#include <vector>
+
+namespace mbed {
+/** \addtogroup drivers */
+
+/** CPU UID reader class
+ *
+ * @note Synchronization level: Interrupt safe
+ * @ingroup drivers
+ */
+class CpuUid {
+public:
+    /** CPU UID array typedef
+     */
+    typedef std::vector<uint8_t> cpu_uid_array_t;
+    
+    /** CpuUid constructor
+     */
+    CpuUid();
```

Comment:
I don't understand this code. A CPU UID is a singular resource in a CPU. This should not be instantiated. Either a singleton or a class with only static methods would be appropriate.

Reply:
the whole idea about that was to have a simple class which can be created, used and destroyed easily, and that it is capable of type conversions via the type cast overloads, so a user can use it like the following:

CpuUid cpuuid;
someclass.SetSerialNumber(cpuuid);

whether this SetSerialNumber function needs std::string, const char* or a vector would not be important.

I thought about implementing it as a singleton ... but then there would always be some getInstance() method which gives me a pointer rather than an object where I can use cast overloads - I could do the dereferencing via macro then, yes ... but I wanted to keep it simple ;)

Reply:
You can implement this paradigm as a class that only contains static data. That should have the benefits you describe without the overhead that exists in the current code.

If you want to optimize further, it's probably reasonable to fetch the UID buffer from hardware each time, rather than keeping it in RAM. There shouldn't be a difference in access latency on most devices, particularly if the C HAL's UID fetch is implemented as a `static inline`.

Reply:
Ok sounds reasonable :)

I guess I have to do a little refactoring....